packages/admin-next/dashboard/src/components/common/infinite-list/infinite-list.tsx

```diff
@@ -0,0 +1,138 @@
+  const { t } = useTranslation()
+
+  const {
+    data,
+    error,
+    fetchNextPage,
+    fetchPreviousPage,
+    hasPreviousPage,
+    hasNextPage,
+    isFetching,
+    isPending,
+  } = useInfiniteQuery({
+    queryKey: queryKey,
+    queryFn: async ({ pageParam = 0 }) => {
+      return await queryFn({
+        limit: pageSize,
+        offset: pageParam,
+      } as TParams)
+    },
+    initialPageParam: 0,
+    maxPages: 5,
+    getNextPageParam: (lastPage) => {
+      const moreItemsExist = lastPage.count > lastPage.offset + lastPage.limit
+      return moreItemsExist ? lastPage.offset + lastPage.limit : undefined
+    },
+    getPreviousPageParam: (firstPage) => {
+      const moreItemsExist = firstPage.offset !== 0
+      console.log(
+        firstPage,
+        moreItemsExist
+          ? Math.max(firstPage.offset - firstPage.limit, 0)
+          : undefined
+      )
+
+      return moreItemsExist
+        ? Math.max(firstPage.offset - firstPage.limit, 0)
+        : undefined
+    },
+    ...queryOptions,
+  })
+
+  const items = useMemo(() => {
+    return data?.pages.flatMap((p) => p[responseKey] as TEntity[]) ?? []
+  }, [data, responseKey])
+
+  const parentRef = useRef<HTMLDivElement>(null)
+  const startObserver = useRef<IntersectionObserver>()
+  const endObserver = useRef<IntersectionObserver>()
+
+  useEffect(() => {
+    if (isPending) {
+      return
+    }
+
+    // Clear the old observers
+    startObserver.current?.disconnect()
+    endObserver.current?.disconnect()
+
+    // Define the new observers after we stop fetching
+    if (!isFetching) {
+      // Define the new observers after paginating
+      startObserver.current = new IntersectionObserver((entries) => {
+        if (entries[0].isIntersecting && hasPreviousPage) {
+          fetchPreviousPage()
+        }
+      })
+
+      endObserver.current = new IntersectionObserver((entries) => {
+        if (entries[0].isIntersecting && hasNextPage) {
+          fetchNextPage()
+        }
+      })
+
+      // Register the new observers to observe the new first and last children
+      startObserver.current?.observe(parentRef.current!.firstChild as Element)
+      endObserver.current?.observe(parentRef.current!.lastChild as Element)
+    }
+  }, [
+    fetchNextPage,
+    fetchPreviousPage,
+    hasNextPage,
+    hasPreviousPage,
+    isFetching,
+    isPending,
+  ])
+
+  useEffect(() => {
+    if (error) {
+      toast.error(error.message)
+    }
+  }, [error])
+
+  if (isPending) {
+    return <Skeleton className="h-[148x] w-full rounded-lg" />
+  }
+
+  return (
+    <div ref={parentRef}>
+      {items &&
+        items.map((item) => <div key={item.id}>{renderItem(item)}</div>)}
+
+      {(isFetching || !hasNextPage) && (
+        <div className="my-4 flex flex-col items-center justify-center">
+          {isFetching && <Spinner className="animate-spin" />}
+          {!hasNextPage && <p className="m-2">{t("general.noMoreData")}</p>}
+        </div>
```

Comment:
We can definitely do a better job at the design (will ask Ludvig), let me see what he thinks. I think being explicit makes it clear that there is no more data vs. maybe users thinking the request is just slow.

Reply:
But isn't that apparent from the spinner being shown when loading new data? 

Just think it looks weird that when you have just a few notifications, you can constantly see:

[Notification]
[No more data]

Reply:
Most of the time people won't see the spinner, since the API is quite fast, so they might not even know that a spinner shows up. Anyway, it's a really minor thing, so I'll let Ludvig decide however he wants it handled, I am happy either way.